pkgs/by-name/ni/niri/package.nix

```diff
@@ -38,56 +47,76 @@ rustPlatform.buildRustPackage rec {
     };
   };
 
+  strictDeps = true;
+
   nativeBuildInputs = [
+    clang
     pkg-config
     rustPlatform.bindgenHook
-    autoPatchelfHook
-    clang
-  ];
-
-  buildInputs = [
-    wayland
-    systemd # For libudev
-    seatd # For libseat
-    libdisplay-info
-    libxkbcommon
-    libinput
-    mesa # For libgbm
-    fontconfig
-    stdenv.cc.cc.lib
-    pipewire
-    pango
   ];
 
-  runtimeDependencies = [
-    wayland
-    mesa
-    libglvnd # For libEGL
-  ];
+  buildInputs =
+    [
+      libdisplay-info
+      libglvnd # For libEGL
+      libinput
+      libxkbcommon
+      mesa # For libgbm
+      pango
+      seatd
+      wayland # For libwayland-client
+    ]
+    ++ lib.optional (withDbus || withScreencastSupport || withSystemd) dbus
+    ++ lib.optional withScreencastSupport pipewire
+    ++ lib.optional withSystemd systemd; # Also includes libudev
 
-  passthru.providedSessions = [ "niri" ];
+  buildFeatures =
+    lib.optional withDbus "dbus"
+    ++ lib.optional withDinit "dinit"
+    ++ lib.optional withScreencastSupport "xdp-gnome-screencast"
+    ++ lib.optional withSystemd "systemd";
+  buildNoDefaultFeatures = true;
 
-  postPatch = ''
-    patchShebangs ./resources/niri-session
-    substituteInPlace ./resources/niri.service \
-      --replace-fail '/usr/bin' "$out/bin"
-  '';
+  postInstall =
+    ''
+      install -Dm0644 resources/niri.desktop -t $out/share/wayland-sessions
+    ''
+    + lib.optionalString withSystemd ''
+      install -Dm0755 resources/niri-session -t $out/bin
+      install -Dm0644 resources/niri{-shutdown.target,.service} -t $out/share/systemd/user
+      install -Dm0644 resources/niri-portals.conf -t $out/share/xdg-desktop-portal
+    '';
 
-  postInstall = ''
-    install -Dm0755 ./resources/niri-session -t $out/bin
-    install -Dm0644 resources/niri.desktop -t $out/share/wayland-sessions
-    install -Dm0644 resources/niri-portals.conf -t $out/share/xdg-desktop-portal
-    install -Dm0644 resources/niri{-shutdown.target,.service} -t $out/share/systemd/user
-  '';
+  env = {
+    # Force linking with libEGL and libwayland-client
+    # so they can be discovered by `dlopen()`
+    RUSTFLAGS = toString (
+      map (arg: "-C link-arg=" + arg) [
+        "-Wl,--push-state,--no-as-needed"
+        "-lEGL"
+        "-lwayland-client"
+        "-Wl,--pop-state"
+      ]
+    );
```

Comment:
These weren't necessary before in nixpkgs? Why are they necessary now? Was this broken before, or is there some other subtle change along with this diff? Was it covered by `runtimeDependencies`, and if so, why isn't that sufficient still? I know i basically didn't even flinch at similar statements in the upstream flake, but uh, i didn't actually understand the nuances of this difference. Why is it aesthetically worse here now?

Reply:
>  Why are they necessary now?

They aren't

>  why isn't that sufficient still?

It worked fine, but it's really overkill for our purposes. `autoPatchelfHook` is for patching pre-built binaries; we don't need to do that when we are compiling. `runtimeDependencies` also links *all* libraries contained in the packages listed, rather than only what's needed (like here). It's sort of like a very big hammer for a very tiny nail

Reply:
ah, yeah that checks out. thanks for the explanation!